benchmark/src/DriverBench/ParallelMultiFileImportBench.php

```diff
@@ -72,61 +74,64 @@ public function beforeIteration(): void
     }
 
     /**
-     * Using Driver's BulkWrite in a single thread
+     * Using Driver's BulkWrite in a single thread.
+     * The number of files to import in each iteration is controlled by the "chunk" parameter.
+     *
+     * @param array{chunk:int} $params
      */
-    public function benchMultiFileImportBulkWrite(): void
+    #[ParamProviders(['provideChunkParams'])]
+    public function benchBulkWrite(array $params): void
     {
-        foreach (self::getFileNames() as $file) {
-            self::importFile($file);
+        foreach (array_chunk(self::getFileNames(), $params['chunk']) as $files) {
+            self::importFile($files);
         }
     }
 
     /**
      * Using library's Collection::insertMany in a single thread
      */
-    public function benchMultiFileImportInsertMany(): void
+    public function benchInsertMany(): void
     {
         $collection = Utils::getCollection();
         foreach (self::getFileNames() as $file) {
             $docs = [];
             // Read file contents into BSON documents
             $fh = fopen($file, 'r');
-            while (($line = fgets($fh)) !== false) {
-                if ($line !== '') {
-                    $docs[] = Document::fromJSON($line);
+            try {
+                while (($line = fgets($fh)) !== false) {
+                    if ($line !== '') {
+                        $docs[] = Document::fromJSON($line);
+                    }
                 }
+            } finally {
+                fclose($fh);
             }
 
-            fclose($fh);
-
             // Insert documents in bulk
             $collection->insertMany($docs);
         }
     }
 
     /**
-     * Using multiple forked threads
+     * Using multiple forked threads. The number of threads is controlled by the "chunk" parameter,
+     * which is the number of files to import in each thread.
      *
-     * @param array{processes:int, files:string[], batchSize:int} $params
+     * @param array{chunk:int} $params
      */
-    #[ParamProviders(['provideProcessesParameter'])]
-    public function benchMultiFileImportFork(array $params): void
+    #[ParamProviders(['provideChunkParams'])]
+    public function benchFork(array $params): void
     {
         $pids = [];
-        foreach (self::getFileNames() as $file) {
-            // Wait for a child process to finish if we have reached the maximum number of processes
-            if (count($pids) >= $params['processes']) {
-                $pid = pcntl_waitpid(-1, $status);
-                unset($pids[$pid]);
-            }
 
+        // Reset to ensure that the existing libmongoc client (via the Manager) is not re-used by the child
+        // process. When the child process constructs a new Manager, the differing PID will result in creation
+        // of a new libmongoc client.
+        Utils::reset();
```

Comment:
Is there existing state prior to `benchFork()` where a Manager would have been reigstered? Perhaps from `beforeIteration()` dropping and creating a collection?

Reply:
Without the reset, I get this error:

> Fatal error: Uncaught MongoDB\Driver\Exception\BulkWriteException: assertion src/mongo/util/hex.cpp:113 in /Users/jerome/Develop/mongo-php-library/benchmark/src/DriverBench/ParallelMultiFileImportBench.php:211

Reply:
Possibly this: https://github.com/mongodb/mongo/blob/r7.0.1/src/mongo/util/hex.cpp#L113

Seems odd that the server would throw back a cryptic assertion failure as its error message, though. If it's easily reproducible this might warrant a bug report in the SERVER project.